README
#
# $Id: README 43 2007-06-21 10:54:39Z sini $
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Susnjar Software Engineering <sini@susnjar.de>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

BIG FAT NOTE

  This module is beta software!!!
  If you find a bug, please report it to me (the author), or if you made an
  improvement, please send the changes (in the form of a context diff) to me,
  so I can integrate it.
  If you would like a feature to be implemented in this module, drop me a line.

INTRODUCTION

  This module implements a perl5 API for accessing CA's AutoSys job control.

  For more information on the AutoSys product itself,
  see CA's website at: http://www3.ca.com/solutions/Product.aspx?ID=253

  This release requires perl5.005 or later, and some release of the AutoSys
  product installed, 4.5 or better should do just fine.

  This package has been known to build with the following combinations
  of software.  Those marked with an asterix (*) are platforms the
  author can not test personally, so there is a remote possibility that
  there may be new bugs in these platforms.  Please report problems to
  the author, of course.

    Operating System          AutoSys Version Perl Version FreeTDS Version
    ========================= =============== ============ ===============
    Cygwin 1.5.23 (WinXP SP2) 4.5             5.8.7        0.64

  NOTE: If you have built it on new platforms, *PLEASE* help save the free
  world by letting me know what version of AutoSys, OS, perl,
  etc. you built it with.

  If you succeed in making this work on any other platform, please send
  the changes (in the form of a context diff) to the author, so I can
  integrate it.

INSTALLATION

  This module installs much like anything else available on CPAN.
  Automatic tests are not yet provided.

    perl Makefile.PL
    make
	make test
    make install
                                                                     
  When using a Sybase backend, you may have to edit your $SYBASE/interfaces file or
  your $HOME/.freetds.conf file and add the appropriate AutoSys DB server(s).

RELEASE NOTES

  The CHANGES file has a complete, historical list of all user-visible
  (and some invisible) changes to this code.

DOCUMENTATION

  POD inside ;-)

AUTHORS

  This module was developed by:

    Sinisa Susnjar <sini@cpan.org>


CHANGES
#
# $Id$
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Sinisa Susnjar <sini@cpan.org>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

Introduction

  This file documents the changes to each and every release of the
  AutoSys perl API, going back as far as the initial 0.99 release.

Changes for 1.04 - release date 29-oct-2007

  o Added status_time variable to CA::AutoSys::Status.
  o Added autorelay example.

Changes for 1.03 - release date 21-jun-2007

  o Replaced dependency DBD::Sybase with the more suitable DBI...
  o Added --dsn option to autorep and sendevent examples.
  o Added a small example web server for controlling AutoSys jobs.
  o Added -q option to autorep example.
  o Cleaned up the DBI mess (a little bit at least).
  o Moved the examples autorep and sendevent to bin directory.
  o Added a small example that shows howto control your AutoSys jobs
    via a web frontend.

Changes for 1.02 - release date 29-mar-2007

  o CA::AutoSys::Status:
    Method format_time() now returns localtime() instead of gmtime() (bug 24962).
    Added new instance variable "name" which contains the long name of the
    appropriate job's status, e.g. "SUCCESS".
  o CA::AutoSys:
    Method new() now accepts the option "dsn" with which one should be able to
    connect to AutoSys databases other than the default (Sybase).
    Note: I have not tested this feature with databases other than Sybase yet,
	since I currently only have access to a Sybase backed AutoSys installation.
	Any feedback is welcome!

Changes for 1.01 - release date 16-jan-2007

  o Added documentation / pods for *.pm files.
  o Added constants for states in CA::AutoSys::Status.

Changes for 1.00 - release date 15-jan-2007

  o Added online help (--help/--man) switches for examples.
  o Added missing META.yml (sorry, it's my first CPAN module...) 

Changes for 0.99 - release date 10-jan-2007

  o Initial development.
  o Implemented basic AutoSys features like job status querying, event sending.


Makefile.PL
#
# $Id: Makefile.PL 45 2007-06-21 11:01:37Z sini $
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Susnjar Software Engineering <sini@susnjar.de>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

use strict;

use ExtUtils::MakeMaker;

WriteMakefile(
	NAME			=> 'CA::AutoSys',
	VERSION_FROM	=> 'lib/CA/AutoSys.pm',
	EXE_FILES		=> [ 'bin/autorep', 'bin/sendevent' ],
	PREREQ_PM		=> { 'DBI' },
	($] >= 5.005 ?
		(ABSTRACT_FROM	=> 'lib/CA/AutoSys.pm',
		 AUTHOR			=> 'Sinisa Susnjar (sini@cpan.org)') : ())
);


lib/CA/AutoSys.pm
#
# $Id: AutoSys.pm 57 2007-10-26 15:10:55Z sini $
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Sinisa Susnjar <sini@cpan.org>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

package CA::AutoSys;

require CA::AutoSys::Job;
require CA::AutoSys::Status;

use strict;
use warnings;
use DBI;

use vars qw($VERSION);

$VERSION = '1.04';

use Exporter;
use vars qw(@ISA @EXPORT);
@ISA = qw(Exporter);
@EXPORT = qw(&new);

our $errstr;

sub new {
	my $self = {};
	my $class = shift();

	$errstr = '';

	if (@_) {
		my %args = @_;
		$self->{dsn} = $args{dsn} ? $args{dsn} : undef;
		$self->{server} = $args{server} ? $args{server} : undef;
		$self->{user} = $args{user} ? $args{user} : undef;
		$self->{password} = $args{password} ? $args{password} : undef;
	}

	if (!defined($self->{dsn})) {
		if (!defined($self->{server})) {
			$errstr = "no dsn given in new()";
			return undef;
		}
		# Default to Sybase when no dsn was given...
		$self->{dsn} = "dbi:Sybase:server=$self->{server}";
	}
	$self->{dbh} = DBI->connect($self->{dsn}, $self->{user}, $self->{password});
	if (!$self->{dbh}) {
		$errstr = "can't connect to dsn ".$self->{dsn}.": ".$DBI::errstr;
		return undef;
	}

	bless($self);
	return $self;
}	# new()

sub find_jobs {
	my $self = shift();
	my $job_name = shift();
	my $job = CA::AutoSys::Job->new(parent => $self, database_handle => $self->{dbh});
	return $job->find_jobs($job_name);
}	# find_jobs()

sub send_event {
	my $self = shift();
	my ($job_name, $event, $status, $event_time);
	if (@_) {
		my %args = @_;
		$job_name = $args{job_name} ? $args{job_name} : '';
		$event = $args{event} ? $args{event} : '';
		$status = $args{status} ? $args{status} : '';
		$event_time = $args{event_time} ? $args{event_time} : '';
	}

	my $sth = $self->{dbh}->prepare(qq{
	exec sendevent '$event', '$job_name', '$status', '', '$event_time', ''
	});

	$sth->execute();

	my ($rc) = $sth->fetchrow_array();
	return $rc;
}	# send_event()

1;
__END__

=head1 NAME

CA::AutoSys - Interface to CA's AutoSys job control.

This module was born out of the need to control some AutoSys jobs via a Perl/CGI script.
It is sort of a quick hack, but it works for me... should you have some wishes / requirements
that are not mentioned in L</TODOs>, please let me know.

=head1 SYNOPSIS

    use CA::AutoSys;

    my $hdl = CA::AutoSys->new( [OPT] ) ;
    my $jobs = $hdl->find_jobs($jobname) ;
    while (my $job = $jobs->next_job()) {
        :
    }
    my $status = $job->get_status() ;
    my $children = $job->find_children() ;
    while (my $child = $children->next_child()) {
        :
    }

=head1 CLASS METHODS

=head2 B<new() >

    my $hdl = CA::AutoSys->new( [OPT] ) ;

Creates a new CA::AutoSys object.

Below is a list of valid options:

=over 5

=item B<dsn>

Specify the DSN of the AutoSys' database server to connect to. If nothing is specified, Sybase will be
assumed: dbi:Sybase:server=<your_server>
With this option you should be able to connect to databases other than Sybase.

=item B<server>

This option is deprecated - rather use the dsn option.
Specify the AutoSys' database server to connect to. Either this option or the dsn option above must be given.
Please note, that when specifying this server option, a Sybase database backend is assumed.

=item B<user>

Specify the database user. With an out-of-the-box AutoSys installation, the default user should work.

=item B<password>

Specify the database password. With an out-of-the-box AutoSys installation, the default password should work.

=back

Example:

    my $hdl = CA::AutoSys->new(server => "AUTOSYS_DEV");

=head1 INSTANCE METHODS

=head2 B<find_jobs() >

    my $jobs = $hdl->find_jobs($jobname) ;

Finds jobs with a given name. When you have the wildcard character '%' somewhere in the job name,
it will return all matching jobs, i.e.:

To find all jobs starting with the string 'MY_JOB':

    $jobs = $hdl->find_jobs('MY_JOB%');

To find all jobs that have the string 'JOB' somewhere in the name:

    $jobs = $hdl->find_jobs('%JOB%');

To find a job with an exact name:

    $jobs = $hdl->find_jobs('JOB_42');

See also L<CA::AutoSys::Job|CA::AutoSys::Job>

=head2 B<send_event() >

    my $rc = $hdl->send_event( [OPT] ) ;

Sends an event to the given job. Returns 1 on success, 0 otherwise.
At least the event name should be given. Depending on the event, more options may be necessary (see below).
For details, consult your AutoSys' User Guide.

Below is a list of valid options:

=over 5

=item B<job_name>

The name of the job - no wildcards allowed.

=item B<event>

Event name. The following list contains all available event names in alphabetical order:

    ALARM                  CHANGE_PRIORITY        CHANGE_STATUS
    CHECK_HEARTBEAT        CHK_BOX_TERM           CHK_MAX_ALARM
    CHK_N_START            CHK_RUN_WINDOW         COMMENT
    DELETEJOB              EXTERNAL_DEPENDENCY    FORCE_STARTJOB
    HEARTBEAT              JOB_OFF_HOLD           JOB_OFF_ICE
    JOB_ON_HOLD            JOB_ON_ICE             KILLJOB
    QUE_RECOVERY           REFRESH_BROKER         RESEND_EXTERNAL_STATUS
    SEND_SIGNAL            SET_GLOBAL             STARTJOB

=item B<status>

The job status when the event is C<CHANGE_STATUS>. The following list contains all possible states for the C<CHANGE_STATUS> event in alphabetical order:

    ACTIVATED              FAILURE                INACTIVE
    ON_HOLD                ON_ICE                 QUE_WAIT
    REFRESH_DEPENDENCIES   REFRESH_FILEWATCHER    RESTART
    RUNNING                STARTING               SUCCESS
    TERMINATED

=item B<event_time>

Use this when you want to schedule an event at a given time.
The argument should have the format 'YYYY/MM/DD HH:MM:SS'.

=back

To force a job start at a given time:

    my $rc = $hdl->send_event(job_name => 'HAPPY_NEW_YEAR', event => 'FORCE_STARTJOB',
                              event_time => '2007/12/31 23:59:59');

To mark an job as inactive:

    my $rc = $hdl->send_event(job_name => 'JOB_42', event => 'CHANGE_STATUS', status => 'INACTIVE');

=head1 TODOs

Make the interface more "perlish", e.g. return an array of jobs instead of forcing the user
to call C<next_job()> / C<next_child()>.

Make the interface more "OO", e.g. allow the user to send an event directly from an C<CA::AutoSys::Job> object
to the underlying job instead of having to use C<CA::AutoSys-E<gt>send_event()>.

There are lots of missing AutoSys features, e.g. "alarms".

=head1 SEE ALSO

L<CA::AutoSys::Job|CA::AutoSys::Job>, L<CA::AutoSys::Status|CA::AutoSys::Status>

=head1 AUTHOR

Sinisa Susnjar <sini@cpan.org>

=head1 MODIFICATION HISTORY

See the CHANGES file.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Sinisa Susnjar. All rights reserved.

This program is free software; you can use and redistribute it under the terms of the L-GPL.
See the LICENSE file for details.


lib/CA/AutoSys/Job.pm
#
# $Id: Job.pm 57 2007-10-26 15:10:55Z sini $
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Sinisa Susnjar <sini@cpan.org>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

package CA::AutoSys::Job;

require CA::AutoSys::Status;

use strict;
use warnings;

use Exporter;
use vars qw(@ISA @EXPORT $VERSION);
@ISA = qw(Exporter);
@EXPORT = qw(&new);

$VERSION = '1.03';

my $debug = 0;

sub new {
	my $self = {};
	my $class = shift();

	if (@_) {
		my %args = @_;
		$self->{parent} = $args{parent} ? $args{parent} : undef;
		$self->{dbh} = $args{database_handle} ? $args{database_handle} : undef;
		$self->{sth} = $args{statement_handle} ? $args{statement_handle} : undef;
		$self->{parent_job} = $args{parent_job} ? $args{parent_job} : undef;
	}

	if ($debug) {
		printf("DEBUG: Job(%s) created.\n", $self);
	}

	bless($self);
	return $self;
}	# new()

sub _fetch_next {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): _fetch_next()\n", $self);
	}
	my ($job_name, $job_type, $joid, $last_start, $last_end, $status, $status_time, $run_num, $ntry, $exit_code, $owner,
		$permission, $date_conditions, $days_of_week, $start_times, $description, $alarm_if_fail, $condition,
		$command, $std_out_file, $std_err_file, $machine, $max_run_alarm, $box_name);

	if (($job_name, $job_type, $joid, $last_start, $last_end, $status, $status_time, $run_num, $ntry, $exit_code, $owner,
		$permission, $date_conditions, $days_of_week, $start_times, $description, $alarm_if_fail, $condition,
		$command, $std_out_file, $std_err_file, $machine, $max_run_alarm, $box_name) = $self->{sth}->fetchrow_array()) {
		if (defined($self->{parent_job})) {
			$self->{parent_job}->{child_cnt}++;
		}
		$condition =~ s/^[ ]*$//g;
		$command =~ s/^[ ]*$//g;
		$machine =~ s/^[ ]*$//g;
		$std_out_file =~ s/^[ ]*$//g;
		$std_err_file =~ s/^[ ]*$//g;
		$self->{job_name} = $job_name;
		$self->{job_type} = $job_type;
		$self->{joid} = $joid;
		$self->{owner} = $owner;
		$self->{permission} = $permission;
		$self->{date_conditions} = $date_conditions;
		$self->{days_of_week} = $days_of_week;
		$self->{start_times} = $start_times;
		$self->{description} = $description;
		$self->{alarm_if_fail} = $alarm_if_fail;
		$self->{condition} = $condition;
		$self->{command} = $command;
		$self->{std_out_file} = $std_out_file;
		$self->{std_err_file} = $std_err_file;
		$self->{machine} = $machine;
		$self->{max_run_alarm} = $max_run_alarm;
		$self->{box_name} = $box_name;
		$self->{status} = CA::AutoSys::Status->new(parent => $self->{parent}, last_start => $last_start,
													last_end => $last_end, status => $status, run_num => $run_num,
													ntry => $ntry, exit_code => $exit_code,
													status_time => $status_time);
		return $self;
	} else {
		$self->{sth}->finish();
		return undef;
	}
}	# _fetch_next()

sub _query {
	my $query = qq{
		select	j.job_name, j.job_type, j.joid, s.last_start, s.last_end, s.status, s.status_time, s.run_num, s.ntry, s.exit_code,
				j.owner, j.permission, j.date_conditions, j.days_of_week, j.start_times, j.description,
				j.alarm_if_fail, j.condition, j.command, j.std_out_file, j.std_err_file, j.machine, j.max_run_alarm,
				j2.job_name as box_name
		from	job j join job_status s
		on		j.joid = s.joid
		left outer join job j2
		on		j.box_joid = j2.joid
	};
	return $query;
}	# _query()

sub find_jobs {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): find_jobs()\n", $self);
	}
	my $job_name = shift();
	my $query = _query() . qq{
		where	j.job_name like '$job_name'
		order by j.joid
	};
	my $sth = $self->{dbh}->prepare($query);
	if (!$sth->execute()) {
		$self->{parent}->{errstr} = "can't select info for job ".$job_name.": ".$sth->errstr();
		return undef;
	}
	return CA::AutoSys::Job->new(parent => $self->{parent}, database_handle => $self->{dbh},
								statement_handle => $sth);
}	# find_jobs()

sub find_children {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): find_children()\n", $self);
	}
	my $query = _query() . qq{
		where	j.box_joid = $self->{joid}
		order by j.joid
	};
	my $sth = $self->{dbh}->prepare($query);
	if ($debug) {
		printf("DEBUG: Job(%s): selecting children for joid %d\n", $self, $self->{joid});
	}
	$self->{child_cnt} = 0;
	if (!$sth->execute()) {
		$self->{parent}->{errstr} = "can't select children for job ".$self->{job_name}.": ".$sth->errstr();
		return undef;
	}
	return CA::AutoSys::Job->new(parent => $self->{parent}, database_handle => $self->{dbh},
								statement_handle => $sth, parent_job => $self);
}	# find_children()

sub next_job {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): next_job()\n", $self);
	}
	return $self->_fetch_next();
}	# next_job()

sub next_child {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): next_child()\n", $self);
	}
	return $self->_fetch_next();
}	# next_child()

sub get_status {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): get_status()\n", $self);
	}
	return $self->{status};
}	# get_status()

sub has_children {
	my $self = shift();
	if ($debug) {
		printf("DEBUG: Job(%s): has_children()\n", $self);
	}
	if (!defined($self->{child_cnt}) || $self->{child_cnt} == 0) {
		return 0;
	}
	return 1;
}	# has_children()

1;

__END__

=head1 NAME

CA::AutoSys::Job - Object representing an AutoSys job.

=head1 INSTANCE METHODS

=head2 B<next_job() >

    my $job = $jobs->next_job() ;

Returns the next job from a list of jobs previously acquired by a call to L<find_jobs()|CA::AutoSys/find_jobs() >.

=head2 B<find_children() >

    my $children = $job->find_children() ;

Returns child jobs for a given job object. The child jobs can be traversed like this:

    my $children = $job->find_children() ;
    while (my $child = $children->next_child()) {
        # do something
        :
    }

=head2 B<next_child() >

    my $child = $children->next_child() ;

Returns the next child from a list of child jobs previously acquired by a call to L<find_children() >.

=head2 B<get_status() >

    my $status = $job_or_child->get_status() ;

Returns a hashref that can be queried for job status variables.
See L<CA::AutoSys::Status|CA::AutoSys::Status> for a list of possible status variables.

=head2 B<has_children() >

    my $rc = $job->has_children() ;

Returns 1 if the given job/child has children, 0 otherwise.

=head1 INSTANCE VARIABLES

=head2 B<job_name>

    print "job_name: ".$job->{job_name}."\n";

Contains the name of the AutoSys job.

=head2 B<job_type>

    print "job_type: ".$job->{job_type}."\n";

Contains the type of the job, c=JOB, b=BOX.

=head2 B<joid>

    print "joid: ".$job->{joid}."\n";

Contains the internal job id in the AutoSys database.

=head1 SEE ALSO

L<CA::AutoSys::Status|CA::AutoSys::Status>, L<CA::AutoSys|CA::AutoSys>

=head1 AUTHOR

Sinisa Susnjar <sini@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Sinisa Susnjar. All rights reserved.

This program is free software; you can use and redistribute it under the terms of the L-GPL.
See the LICENSE file for details.


lib/CA/AutoSys/Status.pm
#
# $Id: Status.pm 40 2007-06-21 10:52:28Z sini $
#
# CA::AutoSys - Perl Interface to CA's AutoSys job control.
# Copyright (c) 2007 Susnjar Software Engineering <sini@susnjar.de>
# See LICENSE for terms of distribution.
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

package CA::AutoSys::Status;

use strict;
use warnings;

use Exporter;
use vars qw(@ISA @EXPORT $VERSION);
@ISA = qw(Exporter);
@EXPORT = qw(&new &format_status &format_time);

$VERSION = '1.03';

our %status_names = (
	0 => '  ',	# *empty*
	1 => 'RU',	# running
	2 => ' 2',	# *not defined*
	3 => 'ST',	# starting
	4 => 'SU',	# success
	5 => 'FA',	# failure
	6 => 'TE',	# terminated
	7 => 'OI',	# on ice
	8 => 'IN',	# inactive
	9 => 'AC',	# activated
	10 => 'RE',	# restart
	11 => 'OH',	# on hold
	12 => 'QW',	# queue wait
	13 => '13',	# *not defined*
	14 => 'RD',	# refresh dependencies
	15 => 'RF',	# refresh filewatcher
);

use constant {
	NONE		=> 0,
	RUNNING		=> 1,
	UNDEF_2		=> 2,
	STARTING	=> 3,
	SUCCESS		=> 4,
	FAILURE		=> 5,
	TERMINATED	=> 6,
	ON_ICE		=> 7,
	INACTIVE	=> 8,
	ACTIVATED	=> 9,
	RESTART		=> 10,
	ON_HOLD		=> 11,
	QUEUE_WAIT	=> 12,
	UNDEF_13	=> 13,
	REFRESH_DEP	=> 14,
	REFRESH_FW	=> 15
};

our %long_status = (
	0		=> "NONE",
	1		=> "RUNNING",
	2		=> "UNDEF_2",
	3		=> "STARTING",
	4		=> "SUCCESS",
	5		=> "FAILURE",
	6		=> "TERMINATED",
	7		=> "ON_ICE",
	8		=> "INACTIVE",
	9		=> "ACTIVATED",
	10		=> "RESTART",
	11		=> "ON_HOLD",
	12		=> "QUEUE_WAIT",
	13		=> "UNDEF_13",
	14		=> "REFRESH_DEP",
	15		=> "REFRESH_FW"
);

sub new {
	my $self = {};
	my $class = shift();

	if (@_) {
		my %args = @_;
		$self->{last_start} = $args{last_start} ? $args{last_start} : undef;
		$self->{last_end} = $args{last_end} ? $args{last_end} : undef;
		$self->{status} = $args{status} ? $args{status} : undef;
		$self->{name} = $long_status{$self->{status}};
		$self->{run_num} = $args{run_num} ? $args{run_num} : undef;
		$self->{ntry} = $args{ntry} ? $args{ntry} : undef;
		$self->{exit_code} = $args{exit_code} ? $args{exit_code} : undef;
	}

	bless($self);
	return $self;
}	# new()

sub format_status {
	my $status = shift();
	return $status_names{$status};
}	# format_status()

sub format_time {
	my $time = shift();
	if (!defined($time) || $time == 999999999) {
		return "-----";
	}
	my ($sec, $min, $hour, $mday, $mon, $year, $wday, $yday) = localtime($time);
	$mon++;
	$year += 1900;
	return sprintf("%02d/%02d/%04d  %02d:%02d:%02d", $mon, $mday, $year, $hour, $min, $sec);
	# return sprintf("%02d/%02d/%04d  %02d:%02d:%02d (%d)", $mon, $mday, $year, $hour, $min, $sec, $time);
}	# format_time()

1;

__END__

=head1 NAME

CA::AutoSys::Status - Object representing an AutoSys job status.

=head1 CLASS METHODS

=head2 B<format_status() >

    my $status_string = $status->format_status($status->{status}) ;

Returns a two character string that represents the status like the AutoSys 'autorep' tool.

=head2 B<format_time() >

    my $time_string = $status->format_time($status->{last_start}) ;

Returns a time string that looks like the one from AutoSys' 'autorep' tool.

=head1 INSTANCE VARIABLES

=head2 B<last_start>

    print "last_start: ".$status->{last_start}."\n";

Contains the time when the job was last started or 999999999 (CA's equivalent for never).
Time is measured in non leap seconds since the epoch, i.e. like the return value of perl's time() function.

=head2 B<last_end>

    print "last_end: ".$status->{last_end}."\n";

Contains the time when the job last ended or 999999999 (CA's equivalent for never).
Time is measured in non leap seconds since the epoch, i.e. like the return value of perl's time() function.

=head2 B<status>

    print "status: ".$status->{status}."\n";

Contains an integer value that represents the status of the job.
The various integer values are mapped to these constants:

    NONE        = 0
    RUNNING     = 1
    UNDEF_2     = 2
    STARTING    = 3
    SUCCESS     = 4
    FAILURE     = 5
    TERMINATED  = 6
    ON_ICE      = 7
    INACTIVE    = 8
    ACTIVATED   = 9
    RESTART     = 10
    ON_HOLD     = 11
    QUEUE_WAIT  = 12
    UNDEF_13    = 13
    REFRESH_DEP = 14
    REFRESH_FW  = 15

=head2 B<name>

    print "status name: ".$status->{name}."\n";

Contains the long name of the appropriate integer value that represents the status of the job.
Can be used instead of the more cryptic output of the L<format_status() > method.
The various integer values are mapped to these constants:

    "NONE"        = 0
    "RUNNING"     = 1
    "UNDEF_2"     = 2
    "STARTING"    = 3
    "SUCCESS"     = 4
    "FAILURE"     = 5
    "TERMINATED"  = 6
    "ON_ICE"      = 7
    "INACTIVE"    = 8
    "ACTIVATED"   = 9
    "RESTART"     = 10
    "ON_HOLD"     = 11
    "QUEUE_WAIT"  = 12
    "UNDEF_13"    = 13
    "REFRESH_DEP" = 14
    "REFRESH_FW"  = 15

=head2 B<run_num>

    print "run_num: ".$status->{run_num}."\n";

Contains an integer value that shows an AutoSys internal run number.
It is currently only used to be output-compatible with CA's 'autorep' tool.

=head2 B<ntry>

    print "ntry: ".$status->{ntry}."\n";

Contains an integer that shows how often the run of the job has been retried by AutoSys.

=head2 B<exit_code>

    print "exit_code: ".$status->{exit_code}."\n";

Contains the last exit code of the job.

=head1 SEE ALSO

L<CA::AutoSys::Job|CA::AutoSys::Job>, L<CA::AutoSys|CA::AutoSys>

=head1 AUTHOR

Sinisa Susnjar <sini@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2007 Sinisa Susnjar. All rights reserved.

This program is free software; you can use and redistribute it under the terms of the L-GPL.
See the LICENSE file for details.
